lexer/lex.go

```diff
@@ -5,9 +5,164 @@ import (
 	"github.com/NeowayLabs/abad/token"
 )
 
-type (
-	Tokval struct {
-		Type  token.Type
-		Value utf16.Str
+type Tokval struct {
+	Type  token.Type
+	Value utf16.Str
+}
+
+func (t Tokval) Equal(other Tokval) bool {
+	return t.Type == other.Type && t.Value.Equal(other.Value)
+}
+
+// Lex will lex the given crappy JS code (utf16 yay) and provide a
+// stream of tokens as a result (the returned channel).
+//
+// The caller should iterate on the given channel until it is
+// closed indicating a EOF (or an error). Errors should be
+// handled by checking the type of the token.
+//
+// A goroutine will be started to lex the given code, if you
+// do not iterate the returned channel the goroutine will leak,
+// you MUST drain the provided channel.
+func Lex(code utf16.Str) <-chan Tokval {
+	tokens := make(chan Tokval)
+	
+	go func() {
+	
+		currentState := lexerInitialState(code)
+		
+		for currentState != nil {
+			token, newState := currentState()
+			if token != nil {
+				tokens <- *token
+			}
+			currentState = newState
+		}
+		
+		close(tokens)
+	}()
+
+	return tokens
+}
+
+type lexerState func() (*Tokval, lexerState)
```

Comment:
Every state returns a token (it could be token.Illegal if a syntax error is found).